### DISCUSSION TITLE
Field Validator for computed_field

### DISCUSSION THREAD

**Discussion by pirnerjonas** [pydantic/pydantic#8865] [Question]
Hi, 

is it somehow possible to use `field_validator` for a `computed_field`? In the following example I want the last example to fail because it doesn't start with "product".

```python
from pydantic import BaseModel, computed_field, field_validator

class FileName(BaseModel):
    file_name: str

    @computed_field
    @property
    def product(self) -> str:
        return self.file_name.split('_')[0]
    
    @computed_field
    @property
    def date(self) -> str:
        return self.file_name.split('_')[1]

    @field_validator('product')
    def name_should_start_with_product(cls, v, values):
        if not v.startswith('Product'):
            raise ValueError('name should start with Product')
        return v
    
file_name = 'Product1_2021-07-01_12-00-00'
f = FileName(file_name=file_name)

# should raise error because name does not start with Product
f2 = FileName(file_name="Articel2_2021-07-01_12-00-00")
f2.model_dump()
```

Thanks for helping :)

**Comment by GasPasChaud** [comment]
Hello,

I think you should validate the `file_name` directly.

Simply change `@field_validator('product')` to `@field_validator('file_name')`

After doing so, here is what I got as output:
```
pydantic_core._pydantic_core.ValidationError: 1 validation error for FileName
file_name
  Value error, name should start with Product [type=value_error, input_value='Articel2_2021-07-01_12-00-00', input_type=str]
    For further information visit https://errors.pydantic.dev/2.6/v/value_error
```

Which is what you wanted right?

**Comment by pirnerjonas** [comment]
Hi,

actually my real example is a little bit more complicated and it would be nice to use the validator directly on the field and not on `file_name` since then I have to define the parsing logic again in the validator.. e.g.:

```python
from pydantic import BaseModel, computed_field, field_validator

class FileName(BaseModel):
    file_name: str

    @computed_field
    @property
    def product(self) -> str:
        return self.file_name.split('_')[0]
    
    @computed_field
    @property
    def date(self) -> str:
        return self.file_name.split('_')[1]

    @computed_field
    @property
    def family(self) -> str:
        return self.file_name.split('_')[1]
    
    @field_validator('file_name')
    def name_should_start_with_product_file_name(cls, v):
        # have to reapeat..
        if not v.split('_')[1].startswith('Family'):
            raise ValueError('name should start with Family')
        return v

    # what i would want:
    @field_validator('family')
    def name_should_start_with_product(cls, v, values):
        if not v.startswith('Family'):
            raise ValueError('name should start with Product')
        return v

```

**Reply by GasPasChaud** [reply]
If all the computed fields can be derived from the single `file_name` field, I would still suggest using a validator for `file_name`. Another way of doing it could be to redefine your model as below and add a `classmethod` to construct from the filename

```
from typing import Self


class Filename(BaseModel):
    file_name: str
    product: str
    family: str
    date: str

    @classmethod
    def from_file_name(cls, file_name: str) -> Self:
        # Could also be regex based validation
        try:
            product, date, family = file_name.split("_")
        except ValueError:
            raise ValueError("Could not split file_name into product, date and family")

        if not product.startswith("Product"):
            raise ValueError("name should start with Product")

        if not family.startswith("Family"):
            raise ValueError("name should start with Family")

        # Add some date validation if needed

        return cls(file_name=file_name, product=product, date=date, family=family)

```

Maybe I'm missing something, but with the provided examples, I do not see the need to validate the computed_field (I don't think that it's possible). If you really don't want to validte the `file_name`, why not perform your validation in the `computed_fields` logic?

_EDIT: sent comment before being finished_

**Reply by pirnerjonas** [reply]
> why not perform your validation in the computed_fields logic

Very valid point.. thats maybe the best idea 😄 .. thanks for the hint

**Reply by GasPasChaud** [reply]
Not that it matters much, but I think that it somewhat is best if you have the validation upon model creation. This way users (or yourself) are guaranteed to have the right filename in their model even without calling the computed fields. 

But it's beyond the point of your original question. I'm glad you found a way to have your code work as intended

**Reply by pirnerjonas** [reply]
sorry for the late reply.. Thanks for mentioning it again! You are absolutely right I ended up taking that approach because it makes much more sense that way 👍
